File: src/Calc/Creation.cpp
/**
 * \file
 * \author Victor von Rhein <goliath@caern.de>
 *
 * \section License
 *
 * Copyright (C) 2011 by Victor von Rhein
 *
 * This file is part of SoulCreator.
 *
 * SoulCreator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SoulCreator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SoulCreator.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <QDebug>

// #include "Exceptions/Exception.h"
#include "Config/Config.h"

#include "Creation.h"


const QList< cv_AbstractTrait::Type > Creation::v_types = QList< cv_AbstractTrait::Type >()
		<< cv_AbstractTrait::Attribute
		<< cv_AbstractTrait::Skill
		<< cv_AbstractTrait::Merit
		<< cv_AbstractTrait::Power;

QList< cv_AbstractTrait::Type > Creation::types() {
	return v_types;
}


Creation::Creation( QObject* parent ): QObject( parent ) {
	character = StorageCharacter::getInstance();
	storage = new StorageTemplate( this );

	v_pointsList = *storage->creationPoints();

	connect( this, SIGNAL( pointsChanged() ), this, SLOT( controlPoints() ) );
	connect( character, SIGNAL( speciesChanged(cv_Species::SpeciesFlag)), this, SLOT( controlPoints() ) );
}

Creation::~Creation() {
	delete storage;
}



cv_CreationPointsList Creation::pointsList() const {
	return v_pointsList;
}

void Creation::setPoints( cv_CreationPointsList pts ) {
	if ( v_pointsList != pts ) {
		v_pointsList = pts;

		emit pointsChanged();
	}
}

void Creation::calcPoints( Trait* trait ) {
	QList< Trait* > list;
	QList< int > pointList;
	QList< cv_AbstractTrait::Category > categories = cv_AbstractTrait::getCategoryList( trait->type() );

	// Für die Fertigkeitsspezialisierungen.
	int dets = 0;
	
	for ( int i = 0; i < categories.count(); ++i ) {
		int pts = 0;

		list = character->traits( trait->type(), categories.at( i ) );

		for ( int j = 0; j < list.count(); ++j ) {
			// Alle Punkte über 4 kosten 2 Erschaffungspunkte
			int ans = list.at( j )->value() - Config::creationTraitDouble;

			if ( ans < 0 ) {
				ans = 0;
			}

			pts += list.at( j )->value() - ans;

			pts += ans * 2;

			if (!list.at(j)->details().isEmpty()) {
				dets += list.at(j)->details().count();
// 				qDebug() << Q_FUNC_INFO << dets;
			}
		}

		pointList.append( pts );
	}

	// Liste wird nach Größe sortiert, damit steht die größte Zahl ganz am Anfang un danach absteigend.
	qSort( pointList );

	// Natürlich gilt das nicht für den Eintrag des Spezialisierungen, die immer am Anfang stehen.
	// Am Anfang deswegen, weil qSort von Klein nach groß sortiert, und ich deswegen von hinten die Liste auslese.
	if (trait->type() == cv_AbstractTrait::Skill){
		pointList.prepend( dets );
	}

	// Bei Attributen ist der jeweils erste Punkt umsonst.
	if ( trait->type() == cv_AbstractTrait::Attribute ) {
		for ( int i = 0; i < pointList.count(); ++i ) {
			v_pointsList.pointList( character->species(), trait->type() )->operator[]( i ) = storage->creationPoints()->pointList( character->species(), trait->type() )->at( i ) + 3 - pointList.at( pointList.count() - 1 - i );

// 			qDebug() << Q_FUNC_INFO << *v_pointsList.pointList(character->species(), trait->type());
		}
	} else
		if ( trait->type() == cv_AbstractTrait::Merit || trait->type() == cv_AbstractTrait::Power ) {
			int sum = 0;

			for ( int i = 0; i < pointList.count(); ++i ) {
				sum += pointList.at( i );
			}

			v_pointsList.pointList( character->species(), trait->type() )->operator[]( 0 ) = storage->creationPoints()->pointList( character->species(), trait->type() )->at( 0 ) - sum;

// 			qDebug() << Q_FUNC_INFO << *v_pointsList.pointList(character->species(), trait->type()) << character->species() << trait->type();
		} else {
			for ( int i = 0; i < pointList.count(); ++i ) {
// 			qDebug() << Q_FUNC_INFO << pointList.count();

				v_pointsList.pointList( character->species(), trait->type() )->operator[]( i ) = storage->creationPoints()->pointList( character->species(), trait->type() )->at( i ) - pointList.at( pointList.count() - 1 - i );

// 			qDebug() << Q_FUNC_INFO << *v_pointsList.pointList(character->species(), trait->type());
			}
		}

	emit pointsChanged();
}


void Creation::controlPoints() {
	for ( int i = 0; i < v_types.count(); ++i ) {
		bool isZero = true;
		bool isNegative = false;

		for ( int j = 0; j < v_pointsList.pointList( character->species(), v_types.at( i ) )->count(); ++j ) {
			if ( v_pointsList.pointList( character->species(), v_types.at( i ) )->at( j ) < 0 ) {
				// Wenn schon eine Kategorie negativ ist, muß ich nicht weiterkontrollieren.
				isNegative = true;
				break;
			} else if ( v_pointsList.pointList( character->species(), v_types.at( i ) )->at( j ) != 0 ) {
				isZero = false;
			}
		}

		if (isNegative){
			emit pointsNegative( v_types.at( i ) );
			continue;
		} else if ( isZero ) {
			emit pointsDepleted( v_types.at( i ) );
		} else {
			emit pointsPositive( v_types.at( i ) );
		}
	}
}

File: src/Calc/Creation.py
# -*- coding: utf-8 -*-

"""
# Copyright

Copyright (C) 2012 by Victor
victor@caern.de

# License

This file is part of SoulCreator.

SoulCreator is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

SoulCreator is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
SoulCreator.  If not, see <http://www.gnu.org/licenses/>.
"""




import copy

from PyQt4.QtCore import pyqtSignal as Signal
from PyQt4.QtCore import QObject

#from src.Error import ErrFileNotOpened
import src.Config as Config
#from src import Error
#from src.Debug import Debug




class Creation(QObject):
	"""
	\brief Berechnet die verbleibenden Punkte bei der Charaktererschaffung.
	"""


	pointsChanged = Signal(object)
	pointsChangedNegative = Signal(str)
	pointsChangedPositive = Signal(str)
	pointsChangedZero = Signal(str)


	def __init__(self, template, character, parent=None):
		super(Creation, self).__init__(parent)

		self.__storage = template
		self.__character = character

		self.__creationPoints = self.__storage.creationPoints
		self.__availablePoints = copy.deepcopy(self.__creationPoints)

		self.pointsChanged.connect(self.controlPoints)
		self.__character.speciesChanged.connect(self.controlPoints)


	@property
	def creationPoints(self):
		return self.__creationPoints


	@property
	def creationPointsAvailable(self):
		return self.__availablePoints


	def calcPoints( self, trait ):
		"""
		Berechnet die noch verfügbaren Punkte.

		\note Beim Abzählen, der bereits verbrauchten Punkte werden nur Eigenschaften beachtet, die auch der aktuellen Spezies des Charakters angehören. Damit können bspw. Punkte für die Kräfte eines Mages vergeben werden, dann kann man die Spezies ändern, frische Punkte für die Kräfte als Changeling vergeben und wieder auf den Mage wechseln und die ursprünglich dort vergebenen Punkte sind noch verfügbar. Beim Speichern allerdings sollten nur die Punkte der aktuellen Spezies beachtet werden, so daß dieser Effekt beim Laden natürlich nicht eintritt.
		"""

		## Herausfinden, welchem Typ diese Eigenschaft angehört.
		typ = None
		stopLoop = False
		for item in self.__storage.traits:
			for subitem in self.__storage.traits[item]:
				if trait.identifier in self.__storage.traits[item][subitem]:
					typ = item
					stopLoop = True
					break
			if stopLoop:
				break

		pointList = []

		categories = self.__storage.categories(typ)

		specialtyPoints = 0
		for item in categories:
			points = 0
			for trait in self.__character.traits[typ][item].values():
				## Es werden nur Eigenschaften beachtet, die auch der aktuellen Spezies des Charakters angehören.
				if not trait.species or trait.species == self.__character.species:
					ans = trait.value - Config.TRAIT_CREATION_DOUBLE_COST

					if ans < 0:
						ans = 0

					points += trait.value - ans
					points += ans * 2

					if trait.specialties:
						specialtyPoints += len(trait.specialties)

			pointList.append(points)

		# Liste wird nach Größe sortiert, damit steht die kleinste Zahl ganz am Anfang und danach aufsteigend.
		pointList.sort()

		# Natürlich gilt das nicht für den Eintrag der Spezialisierungen, die immer am Anfang stehen.
		if typ == "Skill":
			pointList.insert( 0, specialtyPoints )

		# Bei Attributen ist der jeweils erste Punkt umsonst.
		if typ == "Attribute":
			pointList = [item - 3 for item in pointList]
		if typ == "Merit" or typ == "Power":
			ans = sum(pointList)
			pointList = [ans]

		## Attribute, Fertigkeiten und Merits sind für alle Spezies gleich.
		identicalTyps = (
			"Attribute",
			"Skill",
			"Merit",
		)
		if typ in identicalTyps:
			for species in self.__storage.species:
				if self.__storage.species[species]["playable"]:
					self.__availablePoints[species][typ] = [x - y for x, y in zip(self.__creationPoints[species][typ], pointList)]
		else:
			self.__availablePoints[self.__character.species][typ] = [x - y for x, y in zip(self.__creationPoints[self.__character.species][typ], pointList)]
			#Debug.debug("{} {}: {} ({})".format(self.__character.species, typ, self.__availablePoints[self.__character.species], self.__creationPoints[self.__character.species]))

		self.pointsChanged.emit(self.__availablePoints)


	def controlPoints(self):
		"""
		Kontrolliert, ob die Punkte erschöpft sind oder gar übermäßig ausgeschöpft wurden und sendet entsprechende Signale aus.
		"""

		for subitem in self.__availablePoints[self.__character.species]:
			if any(x < 0 for x in self.__availablePoints[self.__character.species][subitem]):
				#Debug.debug("Negativ {}".format(subitem))
				self.pointsChangedNegative.emit(subitem)
			elif any(x > 0 for x in self.__availablePoints[self.__character.species][subitem]):
				#Debug.debug("Positiv {}".format(subitem))
				self.pointsChangedPositive.emit(subitem)
			else:
				#Debug.debug("Zero {}".format(subitem))
				self.pointsChangedZero.emit(subitem)


